### PR TITLE
Uf 8516 flyway and shedlock

### PR DESCRIPTION
**Pull request by mha109**

### CODE REVIEW

File: src/main/java/se/sundsvall/billingdatacollector/service/scheduling/BillingJobHandler.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+	 * Handle the processed jobs
+	 * @param processed The processed jobs
+	 */
+	private void handleProcessed(List<String> processed) {
+		if (processed.isEmpty()) {
+			LOG.info("No flowInstances processed.");
+			return;
+		}
+
+		//TODO handle failed jobs.. something slack.. something email..?
```
Comment (josc04) [inline src/main/java/se/sundsvall/billingdatacollector/service/scheduling/BillingJobHandler.java:58]: Add task for this?

File: src/main/java/se/sundsvall/billingdatacollector/service/scheduling/BillingScheduler.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+	private final BillingJobHandler billingJobHandler;
+
+	public BillingScheduler(BillingJobHandler billingJobHandler) {
+		this.billingJobHandler = billingJobHandler;
+	}
+
+	@Scheduled(cron = "${scheduler.opene.cron.expression}")
+	@SchedulerLock(name = "${scheduler.opene.name}", lockAtMostFor = "${scheduler.opene.lock-at-most-for}")
+	public void runBillingJob() {
+		billingJobHandler.handleBilling();
```
Comment (josc04) [inline src/main/java/se/sundsvall/billingdatacollector/service/scheduling/BillingScheduler.java:20]: Call `RequestId.init()` to initiate it before starting job?

File: src/main/java/se/sundsvall/billingdatacollector/integration/party/PartyClient.java [hunk 3]
```diff
@@ -6,22 +6,24 @@
...
-        path = "/{type}/{legalId}/partyId",
-        produces = { TEXT_PLAIN_VALUE, APPLICATION_PROBLEM_JSON_VALUE }
-    )
-    Optional<String> getPartyId(@PathVariable("type") PartyType partyType, @PathVariable("legalId") String legalId);
+	@GetMapping(
+		path = "/{type}/{legalId}/partyId",
+		produces = { TEXT_PLAIN_VALUE, APPLICATION_PROBLEM_JSON_VALUE }
+	)
+	@Cacheable("partyId")
+	Optional<String> getPartyId(@PathVariable("type") PartyType partyType, @PathVariable("legalId") String legalId);
```
Comment (josc04) [inline src/main/java/se/sundsvall/billingdatacollector/integration/party/PartyClient.java:28]: I'm not sure, but I think that if `Optional` is to be returned when legalId can't be translated (is not found), a `dismiss404 = true` is needed in the `@FeignClient` annotation.

File: src/main/java/se/sundsvall/billingdatacollector/integration/db/model/HistoryEntity.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+	@Column(name = "created")
+	private LocalDate created;
+
+	@Column(name = "location")
+	private String location;
+
+	@PreUpdate
+	@PrePersist
+	public void prePersist() {
+		if(created == null) {
```
Comment (josc04) [inline src/main/java/se/sundsvall/billingdatacollector/integration/db/model/HistoryEntity.java:65]: Is this needed (if `@PreUpdate` is removed)

File: src/main/java/se/sundsvall/billingdatacollector/integration/db/model/ScheduledJobEntity.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+	 * Which end date was used when we fetched data.
+	 */
+	@Column(name = "fetched_end_date", nullable = false)
+	private LocalDate fetchedEndDate;
+
+	/**
+	 * Date when the job was processed.
+	 */
+	@Column(name = "processed", nullable = false)
+	private OffsetDateTime processed;
```
Comment (josc04) [inline src/main/java/se/sundsvall/billingdatacollector/integration/db/model/ScheduledJobEntity.java:50]: Add `@TimeZoneStorage(NORMALIZE)`?

File: src/test/java/se/sundsvall/billingdatacollector/integration/db/model/ScheduledJobEntityTest.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+			.build();
+
+		assertThat(scheduledJobEntity).isNotNull().hasNoNullFieldsOrProperties();
+		assertThat(scheduledJobEntity.getId()).isEqualTo(id);
+		assertThat(scheduledJobEntity.getFetchedStartDate()).isEqualTo(fetchedStartDate);
+		assertThat(scheduledJobEntity.getFetchedEndDate()).isEqualTo(fetchedEndDate);
+		assertThat(scheduledJobEntity.getProcessed()).isEqualTo(processed);
+	}
+
+	@Test
```
Comment (josc04) [inline src/test/java/se/sundsvall/billingdatacollector/integration/db/model/ScheduledJobEntityTest.java:62]: Add test for noDirtOnCreatedBean?
```
assertThat(ScheduledJobEntity.builder().build()).hasAllNullFieldsAndProperties();
assertThat(new ScheduledJobEntity()).hasAllNullFieldsAndProperties();
```

File: src/test/java/se/sundsvall/billingdatacollector/integration/db/model/ScheduledJobEntityTest.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+	void testPrePersist() {
+		ScheduledJobEntity scheduledJobEntity = new ScheduledJobEntity();
+		scheduledJobEntity.prePersist();
+		assertThat(scheduledJobEntity.getProcessed()).isCloseTo(OffsetDateTime.now(), within(1, ChronoUnit.SECONDS));
```
Comment (josc04) [inline src/test/java/se/sundsvall/billingdatacollector/integration/db/model/ScheduledJobEntityTest.java:66]: Add `assertThat(scheduledJobEntity).hasAllNullFieldsAndPropertiesExcept("processed");` to verify that prePersist doesn't modify other attributes than the processed attribute?

File: src/test/java/se/sundsvall/billingdatacollector/integration/db/model/HistoryEntityTest.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+		assertThat(historyEntity).isNotNull().hasNoNullFieldsOrProperties();
+		assertThat(historyEntity.getId()).isEqualTo(id);
+		assertThat(historyEntity.getBillingRecordWrapper()).isEqualTo(billingRecordWrapper);
+		assertThat(historyEntity.getFamilyId()).isEqualTo(familyId);
+		assertThat(historyEntity.getFlowInstanceId()).isEqualTo(flowInstanceId);
+		assertThat(historyEntity.getCreated()).isEqualTo(created);
+		assertThat(historyEntity.getLocation()).isEqualTo(location);
+	}
+
```
Comment (josc04) [inline src/test/java/se/sundsvall/billingdatacollector/integration/db/model/HistoryEntityTest.java:67]: Add test for noDirtOnCreatedBean?
```
assertThat(HistoryEntity.builder().build()).hasAllNullFieldsAndProperties();
assertThat(new HistoryEntity()).hasAllNullFieldsAndProperties();
```

File: src/test/java/se/sundsvall/billingdatacollector/integration/db/model/HistoryEntityTest.java [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+	@Test
+	void testPrepersist() {
+		final var historyEntity = new HistoryEntity();
+		historyEntity.prePersist();
+		assertThat(historyEntity.getCreated()).isCloseTo(LocalDate.now(), within(0, ChronoUnit.DAYS));
```
Comment (josc04) [inline src/test/java/se/sundsvall/billingdatacollector/integration/db/model/HistoryEntityTest.java:72]: Add `assertThat(historyEntity).hasAllNullFieldsAndPropertiesExcept("created");` to verify that prePersist doesn't modify other attributes than the created attribute?

File: src/test/java/se/sundsvall/billingdatacollector/service/DbServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+		verify(mockResponseEntity).getHeaders();
+		verify(mockHistoryRepository).saveAndFlush(argThat(entity ->
+			entity.getFlowInstanceId().equals(FLOW_INSTANCE_ID) &&
+			entity.getFamilyId().equals(FAMILY_ID) &&
+			entity.getLocation().equals("/location/uuid")));
+		verifyNoMoreInteractions(mockFalloutRepository);
+	}
+
+	@Test
+	void testHasAlreadyBeenProcessed_shouldReturnTrue_whenEitherIsTrue() {
```
Comment (josc04) [inline src/test/java/se/sundsvall/billingdatacollector/service/DbServiceTest.java:154]: Consider making `testHasAlreadyBeenProcessed_shouldReturnTrue_whenEitherIsTrue` to a parameterized test (along with `testHasAlreadyBeenProcessed_shouldReturnFalse_whenNeitherIsTrue`) to verify all possible paths. Right now only path when both a true or if existsByFamilyIdAndFlowInstanceId is true are tested. Path where only existsByFamilyIdAndFlowInstanceId is true is never tested.

File: src/test/resources/application-junit.yml [hunk 11]
```diff
@@ -53,13 +53,16 @@ integration:
         - channelA
         - channelB
         - channelC
+scheduler:
+  opene:
+    cron:
+      expression: '0 * * * * *'
```
Comment (josc04) [inline src/test/resources/application-junit.yml:59]: Should the junit profile execute the scheduled job every minute, or can it be tured of (not sure if the scheduled job might interfere with for example db tests if it triggers at the same time as test is executing)?

File: src/main/java/se/sundsvall/billingdatacollector/service/scheduling/BillingJobHandler.java (hunk 1 above)
Comment (mha109) [reply src/main/java/se/sundsvall/billingdatacollector/service/scheduling/BillingJobHandler.java:58]: Already done, UF-8747.

File: src/main/java/se/sundsvall/billingdatacollector/integration/db/model/HistoryEntity.java [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+	@Column(name = "flow_instance_id")
+	private String flowInstanceId;
+
+	@Column(name = "created")
+	private LocalDate created;
+
+	@Column(name = "location")
+	private String location;
+
+	@PreUpdate
```
[the remark this answers is not in the document]
Comment (mha109) [reply src/main/java/se/sundsvall/billingdatacollector/integration/db/model/HistoryEntity.java:62]: Nope, remnant from when I was figuring things out.

File: src/main/java/se/sundsvall/billingdatacollector/service/CollectorService.java [hunk 13]
```diff
@@ -53,81 +54,99 @@ public void trigger(String flowInstanceId) {
...
+								idsToReturn.add(flowInstanceId);
+								triggerBilling(flowInstanceId);
 							} catch (Exception e) {
 								LOG.warn("Failed to trigger billing for familyId: {} and flowInstanceId: {}", supportedFamilyId, flowInstanceId, e);
 							}
-						});
+						} else {
+							LOG.info("Billing for familyId: {} and flowInstanceId: {} has already been processed", supportedFamilyId, flowInstanceId);
+						}
+					});
```
[the remark this answers is not in the document]
Comment (mha109) [reply src/main/java/se/sundsvall/billingdatacollector/service/CollectorService.java:110]: Only if you come up with a name for the method ;)

File: src/main/java/se/sundsvall/billingdatacollector/service/CollectorService.java [hunk 14]
```diff
@@ -53,81 +54,99 @@ public void trigger(String flowInstanceId) {
...
-			preprocessorIntegration.createBillingRecord(billingRecordWrapper.getBillingRecord());
-			LOG.info("Successfully sent record to preprocessor for flowInstanceId: {}", flowInstanceId);
+			var response = preprocessorIntegration.createBillingRecord(billingRecordWrapper.getBillingRecord());
+			LOG.info("Successfully sent record to preprocessor for flowInstanceId: {}", billingRecordWrapper.getFlowInstanceId());
+			dbService.saveToHistory(billingRecordWrapper, response);
 		} catch (Exception e) {
 			//Save the BillingRecordWrapper if we failed to send it to the preprocessor
-			LOG.warn("Failed to create a record for flowInstanceId: {}", flowInstanceId, e);
+			LOG.warn("Failed to create a record for flowInstanceId: {}", billingRecordWrapper.getFlowInstanceId(), e);
 			Optional.of(billingRecordWrapper)
```
[the remark this answers is not in the document]
Comment (mha109) [reply src/main/java/se/sundsvall/billingdatacollector/service/CollectorService.java:70]: Yep, some kind of tankevurpa here.

File: src/test/resources/application-junit.yml (hunk 11 above)
Comment (mha109) [reply src/test/resources/application-junit.yml:59]: It shoiuld be turned off.

File: src/main/java/se/sundsvall/billingdatacollector/integration/db/model/ScheduledJobEntity.java (hunk 5 above)
Comment (mha109) [reply src/main/java/se/sundsvall/billingdatacollector/integration/db/model/ScheduledJobEntity.java:50]: Oops, missed it.

File: src/main/java/se/sundsvall/billingdatacollector/integration/db/model/HistoryEntity.java (hunk 4 above)
Comment (mha109) [reply src/main/java/se/sundsvall/billingdatacollector/integration/db/model/HistoryEntity.java:65]: Nope, it was only there for the ´@PreUpdate´ to not overwrite it, i.e. not needed.

File: src/main/java/se/sundsvall/billingdatacollector/integration/billingpreprocessor/BillingPreprocessorIntegration.java
[the remark this answers is not in the document]
Comment (mha109) [reply src/main/java/se/sundsvall/billingdatacollector/integration/billingpreprocessor/BillingPreprocessorIntegration.java:1]: Probably not, will remove.

File: src/test/java/se/sundsvall/billingdatacollector/service/DbServiceTest.java (hunk 10 above)
Comment (mha109) [reply src/test/java/se/sundsvall/billingdatacollector/service/DbServiceTest.java:154]: Done!

**Review by josc04 — APPROVED** [review]

**Pull request merged**